### ISSUE TITLE
Renaming ThisDocument breaks macrobuttons

### ISSUE THREAD

**Issue by Stevefb** [rubberduck-vba/Rubberduck#3265]
RD makes it easy for users to rename "ThisDocument". Whilst most things continue to work doing so can break some functionality, so far Macrobuttons...

See this question SO: https://stackoverflow.com/questions/36950856/will-renaming-visios-thisdocument-module-cause-any-problems/45705251#45705251

Also, applies to subs with special names within the document module:

- close
- new
- open 
- etc.

![capture](https://user-images.githubusercontent.com/9349975/29390718-80f60762-8335-11e7-8354-58890d7e461f.PNG)

**Comment by retailcoder** [comment]
It would be nice if, when #3252 `ShadowedDeclarationInspection` is merged, the input validation used that inspection fail the validation if the inspection returns a result, given an identifier. Perhaps the inspection could implement some `IShadowedDeclarationFinder` interface that guarantees a `public bool IsShadowingAccessibleIdentifier(string name)` method, which could probably easily be extracted from the existing inspection logic...

**Comment by ThunderFrame** [comment]
@Stevefb Are you perhaps confusing `ThisDocument` and `Document`? `ThisDocument` is the default name of the document component, and it can be changed by the user. `Document` is the name of the implicit `WithEvents` object in the module (and it's name can't be changed by the user), to which event handlers are implemented. Renaming `ThisDocument` to `fooDoc` shouldn't have any impact on the `Document` event handlers like `Document_Open`. Or have you seen otherwise?

Renaming `ThisDocument` to `Document` would be utter madness.

Also, don't forget that the members of word template are inherited by the word document. If your template has a document component with a name of `ThisDocument` and your Word document has a document component with a name of `fooDoc` name, then from within the Word document, you can use both `Debug.Print ThisDocument.Name` **and** `Debug.Print fooDoc.Name`

**Comment by Stevefb** [comment]
@ThunderFrame You are correct, the event handlers for Document_xyz continue to work... what I was getting at was the fact that RD allows me to rename subs in the 'ThisDocument' module (although confusingly I did say 'document module' because the names of those Subs are prefixed with: Document_).

That said, I can 100% confirm that if I change ThisDocument to FooBar (or even Document) things break, so far Macrobuttons break completely.

Utter madness follows... 

![capture](https://user-images.githubusercontent.com/9349975/29412067-91bb5c54-839a-11e7-8fd3-04819d84c594.PNG)

**Comment by retailcoder** [comment]
Macro buttons need to be re-mapped, is all. More of an enhancement than a bug really.

**Comment by ThunderFrame** [comment]
Except, remapping the Macro button field, or even deleting it and adding a new one , does not work.

**Comment by retailcoder** [comment]
Oh. Then we should disable renaming the host document object. Shouldn't be too hard =)

**Comment by ThunderFrame** [comment]
I'm not entirely sure *how* this works, but it seems to fix *all* of the MACROBUTTON fields *after* a document rename, and doesn't require editing or deleting the existing fields.

```vb
On Error Resume Next
Application.Run vbNullString
```

Kind of obvious, huh?

And for Visio, the double-click behavior macro needs to be explicitly reset with something like:

```vb
  Application.Windows.ItemEx("Drawing1:Page-1:Decision <SHAPE>").Activate
  Application.ActiveWindow.Shape.CellsSRC(visSectionObject, visRowEvent, visEvtCellDblClick).FormulaU = "RUNMACRO(""newDocumentName.aaaa"")"
```
